### PR TITLE
SJQ2T-427: BE - Grant permissions to a file

### PR DESCRIPTION
**Pull request by DraganMitrasinovicHtec**
Granting permissions for a file, if the user you are giving the permission to doesn't have access to the bucket/parent folder it will be granted to them.

### CODE REVIEW

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 1]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+
+    AccessTypeEntity accessType =
+        accessTypeRepository
+            .findByName(permissionGrantRequest.getAccess())
+            .orElseThrow(AccessTypeNotFoundException::new);
+
+    FileMetadataEntity file =
+        fileMetadataRepository
+            .findById(permissionGrantRequest.getObjectId())
+            .orElseThrow(BucketNotFoundException::new);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: FileNotFound

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 2]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+    }
+
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessType);
+    permission.setObject(file);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    currentUserHasPermissionForBucket(permissionGrantRequest.getObjectId(), AccessType.OWNER);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: should be `currentUserHasPermissionForFile()`, which check permissions for the file, and also everything on the file's path

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 3]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessTypeView);
+    permission.setObject(parent);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    permissionRepository.save(permission);
+
+    log.info(
+        "User {} ({}) granted {} permission for file {} ({}) to user {} ({})",
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: folder

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 4]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+    return modelMapper.map(savedPermission, PermissionResponse.class);
+  }
+
+  private void grantViewPermissionToAllFoldersUpstream(
+      FolderEntity parent, UserEntity currentUser, UserEntity targetUser) {
+
+    if (parent == null
+        || permissionRepository.existsById(
+            new UserObjectAccessKey(targetUser.getId(), parent.getId()))) {
+      return;
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:352]: is this method supposed to be recursive?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 5]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+      throw new PermissionAlreadyExistsException();
+    }
+
+    UserObjectAccessEntity savedPermission = permissionRepository.save(permission);
+
+    grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);
+
+    permissionGrantRequest.setObjectId(file.getBucket().getId());
+    permissionGrantRequest.setAccess("view");
+    grantPermissionForBucket(permissionGrantRequest);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]:
```
  grantViewPermissionToAllFoldersUpstream(file.getParentFolder(), currentUser, targetUser);

  permissionGrantRequest.setObjectId(file.getBucket().getId());
  permissionGrantRequest.setAccess("view");
  grantPermissionForBucket(permissionGrantRequest);
```
this chunk could probably be a method, grant view permission to upstream (objects)

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 6]
```diff
@@ -237,6 +236,104 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+        permission.getAccess().getName(),
+        permission.getObject().getName(),
+        permission.getObject().getId(),
+        permission.getGrantedTo().getEmail(),
+        permission.getGrantedTo().getId());
+
+    return modelMapper.map(savedPermission, PermissionResponse.class);
+  }
+
+  private void grantViewPermissionToAllFoldersUpstream(
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:346]: right now, this method  is basically grant permission for folder

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java (hunk 5 above)
Comment (ivanfhtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: also, won't `grantPermissionForBucket()` throw an error if the user already has permission for the bucket?

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java (hunk 4 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:352]: Yeah it is, my naming is kinda bad tho

Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:352]: Wow, did I legit forget to call it at the end, that was the whole point.

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java (hunk 5 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java]: 1) Yeah imma extract that
2) Wanted to reuse the method, forgot about that. I think it's easiest to surround it with a try catch

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java (hunk 6 above)
Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:346]: Forgot to write the call to itself :D

File: backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java [hunk 7]
```diff
@@ -237,6 +238,117 @@ public void revokePermission(UUID bucketId, UUID userId) {
...
+
+    UserObjectAccessEntity permission = new UserObjectAccessEntity();
+    permission.setGrantedBy(currentUser);
+    permission.setGrantedTo(targetUser);
+    permission.setAccess(accessTypeView);
+    permission.setObject(parent);
+    permission.setId(
+        new UserObjectAccessKey(permission.getGrantedTo().getId(), permission.getObject().getId()));
+
+    permissionRepository.save(permission);
```
Comment (ivanfhtec) [inline backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:368]: what if the current user doesn't have permission to do this?
should probably be fine though

Comment (DraganMitrasinovicHtec) [reply backend/src/main/java/com/htecgroup/skynest/service/impl/PermissionServiceImpl.java:368]: I'm assuming this method will only be called from inside this service, after the permission check

**Review by DarkoVeljanoskiHtec — APPROVED** [review]
Are You Ready ?

**Comment by DraganMitrasinovicHtec** [comment]
> Are You Ready ?

I was born ready

**Pull request merged**